### DISCUSSION TITLE
Request for improvement in inpaint function

### DISCUSSION THREAD

**Discussion by watdomo** [glucauze/sd-webui-faceswaplab#82] [Ideas]
Thank you from the bottom of my heart for providing us with a great extension.

Inpainting must always be done at a size of 512x512, no matter what the original image size is, or the result will be corrupted.
Perhaps the current inpainting is always done at the image size, so if the original image size is larger than 512x512, the result will be deformed beyond the learned range.
Especially when Lora is specified at the prompt with an image size of 1024 or larger, it produces a noticeably worse image than expected.
This is the same behavior as when inpainting with img2img's inpaint function for sizes larger than 512.

It would be perfect if you could fix the inpaint width and height to 512x512 no matter what size the source image is, like ADetailer does.

Therefore, currently, using this inpainting function will produce unexpected results when using img2img for large size finishing, when using Hires.fix to improve quality, or when generating images in non-square sizes such as 512x768.

Pre-Inpainting (before swapping), Post-Inpainting (After swapping), and Global-Inpainting (all faces) all give the same result.

--
It would be more complete if the size of Inpaint mask blur and Inpaint only masked padding, pixels could also be changed for the ADetailer.

(Example Results)
![inpaintex](https://github.com/glucauze/sd-webui-faceswaplab/assets/111408083/9534b44a-e2e3-4a92-95a2-8f0561040a29)

**Comment by glucauze** [comment]
Hi,

This is the configuration for inpainting i am using right now in https://github.com/glucauze/sd-webui-faceswaplab/blob/main/scripts/faceswaplab_inpainting/i2i_pp.py :

```python
                i2i_kwargs = {
                    "init_images" : [img],
                    "sampler_name": options.inpainting_sampler,
                    "do_not_save_samples": True,
                    "steps": options.inpainting_steps,
                    "width": img.width,
                    "inpainting_fill": 1,
                    "inpaint_full_res": True,
                    "height": img.height,
                    "mask": mask,
                    "prompt": prompt,
                    "negative_prompt": negative_prompt,
                    "denoising_strength": options.inpainting_denoising_strengh,
                    "seed": options.inpainting_seed,
                }
```

It uses inpaint "only masked" and the same resolution as the input image. I think it should keep the resolution unless i am wrong.

I could set the inpainting with 512x512 (or a ratio of it) but that would require reupscaling the image latter after that. Is that what you are expecting ?

**Reply by watdomo** [reply]
Incorporating the insights offered by ADetailer could prove beneficial. A default inpainting resolution of 512x512 would be optimal, while still allowing users the flexibility to modify this setting.

As for SD1.5, it's generally accepted that training occurs on 512x512 material. Consequently, it's widely understood that increasing the resolution could destabilize the generated outcome. This is akin to the notion that, unless working with SDXL, one wouldn't arbitrarily generate images larger than 1024.

Typically, the mask area constitutes a segment of the full image. When this segment is 512, there's usually no need for later upscaling. The primary concern should be accurate prompt reflection in the inpainting process, rather than pushing for higher resolution and risking unforeseen results.

To enhance quality, it makes sense to upscale the entire image using a tiling technique like Ultimate SD Upscale, but only after achieving the desired inpainting output.

**Reply by watdomo** [reply]
The argument implies that having a low denoising strength, such as 0.2, for inpainting isn't a big issue, making either setting acceptable. However, due to current specifications, there's a limit to how much the denoising strength can be increased.

The newly-added inpainting feature is promising for manipulating facial details and expressions. Unfortunately, its application becomes limited when used at large resolutions and high denoising strengths, leading to poor results as noted in the initial post. This proposal aims to navigate through that challenge.

**Comment by glucauze** [comment]
Ok thx i don't have the time to really implement that properly right now. I need to think of all implication of that change. I will try to see when i have time and compare with adetailler.
